src/Core/Content/Seo/SeoUrlPlaceholderHandler.php

```diff
@@ -90,43 +87,55 @@ private function createSeoMapping(SalesChannelContext $context, array $mapping):
             return [];
         }
 
-        $query = new QueryBuilder($this->connection);
-        $query->setTitle('seo_url::replacement');
-        $query->addSelect(['seo_path_info', 'path_info', 'sales_channel_id']);
+        $subqueryParts = '';
+        $params = [];
 
-        $query->from('seo_url');
-        $query->andWhere('seo_url.is_canonical = 1');
-        $query->andWhere('seo_url.path_info IN (:pathInfo)');
-        $query->andWhere('seo_url.language_id = :languageId');
-        $query->andWhere('seo_url.sales_channel_id = :salesChannelId OR seo_url.sales_channel_id IS NULL');
-        $query->andWhere('seo_url.is_deleted = 0');
-        $query->setParameter('pathInfo', $mapping, ArrayParameterType::STRING);
-        $query->setParameter('languageId', Uuid::fromHexToBytes($context->getContext()->getLanguageId()));
-        $query->setParameter('salesChannelId', Uuid::fromHexToBytes($context->getSalesChannelId()));
-
-        $seoUrls = $query->executeQuery()->fetchAllAssociative();
+        foreach ($mapping as $key => $value) {
+            $subqueryParts .= '(SELECT id FROM seo_url WHERE path_info = ? AND language_id = 0x' . $context->getLanguageId() . ' AND is_canonical = 1 AND is_deleted = 0 AND sales_channel_id = 0x' . $context->getSalesChannelId() . ' LIMIT 1) UNION ALL ';
+            $params[] = $value;
+        }
 
-        $mapped = [];
+        $seoUrls = $this->connection->fetchAllKeyValue($this->getFinalQuery($subqueryParts), $params);
 
-        foreach ($seoUrls as $seoUrl) {
-            $key = self::DOMAIN_PLACEHOLDER . $seoUrl['path_info'] . '#';
+        $missing = [];
 
-            // prefer sales channel specific seo urls
-            if ($seoUrl['sales_channel_id'] === null && isset($mapped[$key])) {
-                continue;
+        foreach ($mapping as $key => $value) {
+            if (isset($seoUrls[$value])) {
+                $mapping[$key] = $seoUrls[$value];
+            } else {
+                $missing[$key] = $value;
             }
+        }
 
-            $seoPathInfo = trim((string) $seoUrl['seo_path_info']);
-            if ($seoPathInfo === '') {
-                continue;
-            }
-            $mapping[$key] = $seoPathInfo;
-            $mapped[$key] = true;
+        if (empty($missing)) {
+            return $mapping;
+        }
+
+        $subqueryParts = '';
+        $params = [];
+
+        foreach ($missing as $key => $value) {
+            $subqueryParts .= '(SELECT id FROM seo_url WHERE path_info = ? AND language_id = 0x' . $context->getLanguageId() . ' AND is_canonical = 1 AND is_deleted = 0 AND sales_channel_id IS NULL LIMIT 1) UNION ALL ';
```

Comment:
Could we extract the build sub-query parts to query as a private method so we don't need to repeat it?

Reply:
I may not understand your change-request correctly, but the two queries in the "foreach" loops are not the same, so we can't just create a single method for them.